static/js/services/generate-search-requests.js

```diff
@@ -139,40 +145,54 @@ var _ = require('underscore'),
           return requests;
         },
         contacts: function(filters) {
-          var typeViews = documentType(filters, 'medic-client/contacts_by_type');
-          var freetextViews = freetext(filters, 'medic-client/contacts_by_freetext');
-          if (typeViews && typeViews.params.keys.length &&
-              freetextViews && freetextViews.length) {
-            return freetextViews.map(function(freetextView) {
+          var typeRequests = documentTypeRequest(filters, 'medic-client/contacts_by_type');
+          // Single request, with one key for each type.
+          var hasTypeRequests = typeRequests && typeRequests.params.keys.length;
+
+          var freetextRequests = freetextRequest(filters, 'medic-client/contacts_by_freetext');
+          // One request per word.
+          var hasFreetextRequests = freetextRequests && freetextRequests.length;
+
+          // If both type and freetext requests,
+          // then we make a special combined contacts+type request.
+          if (hasTypeRequests && hasFreetextRequests) {
```

Comment:
After a small renaming, the code is now clear enough to me that I removed all the comments! 🎊 

Hopefully it's not too bad for either of our brains now. Any more thoughts about it?